### PR TITLE
feat/shared_state: ensure mutations to SharedState are approved by the section

### PR DESCRIPTION
**Pull request by maqi**
Close https://github.com/maidsafe/routing/issues/2119

### CODE REVIEW

File: src/section/shared_state.rs [hunk 1]
```diff
@@ -60,11 +62,19 @@ impl SharedState {
             );
         }
 
-        if self.our_history.len() > 1 && *self != new {
+        if self.our_history.len() > 1 {
+            if *self != new {
```
Comment (dirvine) [inline src/section/shared_state.rs:66]: I really want to get this log_or_panic gone ASAP. In this case here will it matter? I mean, it's an Error and a valid Error, why log or panic, if we want to test this we can test for that error. 
If these are semilattice then perhaps we can be even more smart?

Comment (maqi) [reply src/section/shared_state.rs:66]: was just following the previous implementation.
I think the idea here is that shall not be happen, and if happens during test, we want it to be exposed immediately and explicitly.

However, I think there is missing `return` for these `error branches` ? @madadam

Comment (dirvine) [reply src/section/shared_state.rs:66]: Yes, I understand but let's get to launch now and no more log or panic. We don't panic unless we have an unrecoverable error. I am not sure we have that here so we should return the error or fix the error. If we return it we should log Error and if we fix it then perhaps log Debug? 
Panic in tests would mean we did neither and the production code has to do one of these things (return or fix). I say now is the time to stop using log_or_panic and getting rid of it when we are working on those bits of code. 

It's a better cleaner solution to not use this from this point forward.  :+1:

Comment (maqi) [reply src/section/shared_state.rs:66]: ok, I will change it to log and return with an error

Comment (dirvine) [reply src/section/shared_state.rs:66]: Or just `?` if it can be handled upstream.

File: src/section/shared_state.rs [hunk 2]
```diff
@@ -52,25 +55,28 @@ impl SharedState {
...
-            );
+        if self.our_history.len() > 1 {
+            if *self != new {
+                error!(
+                    "shared state update - mismatch: old: {:?} --- new: {:?}",
+                    self, new
+                );
+                return Err(RoutingError::InvalidState);
+            }
+        } else if !self.our_history.self_verify() {
```
Comment (madadam) [inline src/section/shared_state.rs:72]: Two points:
1. You should do the check on `new`, not `self`. `self` is presumably valid, because we only allow approved mutations. But we don't know whether `new` is, because we just received it from someone.
2. You should verify everything, not just `our_history`.

File: src/section/shared_state.rs [hunk 3]
```diff
@@ -173,15 +179,32 @@ impl SharedState {
         // FIXME: we should perform these checks before voting, but once the vote accumulates we
         // must obey it:
         if !self.our_prefix().matches(p2p_node.name()) {
-            trace!("not adding node {} - not matching our prefix", p2p_node);
+            trace!(
+                "not adding node {} - not matching our prefix",
+                p2p_node.name()
```
Comment (madadam) [inline src/section/shared_state.rs:184]: Why this change?

File: src/section/shared_state.rs [hunk 4]
```diff
@@ -173,15 +179,32 @@ impl SharedState {
...
         }
 
         if self.our_members.contains(p2p_node.name()) {
-            trace!("not adding node {} - already a member", p2p_node);
+            trace!("not adding node {} - already a member", p2p_node.name());
             return false;
         }
 
+        // The section public key of the proof shall be trusted by us.
+        match self.our_history.check_trust(iter::once(&proof.public_key)) {
```
Comment (madadam) [inline src/section/shared_state.rs:195]: So this checks that the proof key is known to us, but we should also check the proof itself is valid by verifying the signature. This applies to other places too.

Comment (madadam) [reply src/section/shared_state.rs:195]: Also, `check_trust` isn't supposed to be used like this, although it might still happen to work. What `check_trust` does, is: given a collection of keys that we already trust, return whether the proof chain itself is trusted. So you would use it to verify the proof chain itself, not the keys. But we don't need to verify the proof chain here, we can assume it is valid already. What you should use is just `self.our_history.has_key(&proof.public_key)`.

File: src/section/shared_state.rs [hunk 5]
```diff
@@ -264,6 +301,22 @@ impl SharedState {
...
+            result => {
+                trace!(
+                    "not updating our section {:?} - Old section key {:?} is not trusted.",
+                    result,
+                    section_key.proof.public_key
+                );
+                return;
+            }
+        }
+
```
Comment (madadam) [inline src/section/shared_state.rs:319]: Again, just use `self.our_history.has_key(&section_key.proof.public_key)`. But also verify the signatures and also verify `elders_info`, not just `section_key`. It might be useful to add a `verify` method to `Proven`, to avoid duplication.

File: src/section/shared_state.rs (hunk 2 above)
Comment (maqi) [reply src/section/shared_state.rs:72]: > You should do the check on new

Ah, right. 

> You should verify everything, not just our_history

That means a check against the serialised info itself? I thought such check shall be carried out inside the message accumulation already?
Regarding checking each field individually, I am not sure for each field, which value is `valid`.
For example, `handled_genesis_event` shall be `false` or `true` for new? and `our_members` or `relocate_queue` shall be what? etc.

File: src/section/shared_state.rs (hunk 3 above)
Comment (maqi) [reply src/section/shared_state.rs:184]: just to make the log message cleaner and consistent with outputs in other places

File: src/section/shared_state.rs (hunk 4 above)
Comment (maqi) [reply src/section/shared_state.rs:195]: > but we should also check the proof itself is valid by verifying the signature. This applies to other places too

I thought this has been carried out during message accumulation already?

> What you should use is just self.our_history.has_key(&proof.public_key)

Ok, will change to this

File: src/section/shared_state.rs (hunk 5 above)
Comment (maqi) [reply src/section/shared_state.rs:319]: > But also verify the signatures and also verify elders_info, not just section_key. It might be useful to add a verify method to Proven, to avoid duplication.

I thought such signature check has been carried out already when accumulation. Will have a check through.

File: src/section/shared_state.rs (hunk 2 above)
Comment (madadam) [reply src/section/shared_state.rs:72]: > I thought such check shall be carried out inside the message accumulation already

No, the idea is that any node can send us a message containing a `SharedState` and we should be able to merge it with outs if it is valid. There will be no accumulation for this, because all the info contained in `SharedState` is already signed by the section quorum. Currently this isn't necessary, because we only call `update` in response to accumulated `Genesis` event, but that will eventually change.

Comment (madadam) [reply src/section/shared_state.rs:72]: > For example, handled_genesis_event shall be false or true for new? and our_members or relocate_queue shall be what? etc.

Right, so not everything is ready for parsec removal yet, but we can do at least those parts that are. So:
- `handled_genesis_event` can be ignored (it will always be `false` for incoming shared state)
- `relocate_queue` can be ignored too, for now, at least until we figure out what to do with it (we have a separate issue for it and I think you started looking into it?)
- `our_members` can be verified normally, because the entries in it (`MemberInfo`) all have proofs attached.
- Same with `SectionMap`

File: src/section/shared_state.rs (hunk 3 above)
Comment (madadam) [reply src/section/shared_state.rs:184]: OK. I think it might sometimes be useful to see the full `P2pNode` in the log, not just the name, but it's probably fine here.

File: src/section/shared_state.rs (hunk 2 above)
Comment (maqi) [reply src/section/shared_state.rs:72]: Ok, will do the checks.

>it will always be false for incoming shared state

Then maybe just check the incoming `new`'s `handled_genesis_event ` shall be `false` ?

File: src/section/shared_state.rs (hunk 4 above)
Comment (madadam) [reply src/section/shared_state.rs:195]: > I thought this has been carried out during message accumulation already?

Yeah that's true. Hm. Then maybe it's not necessary to do those checks here. Only where we know the input might come from untrusted sources (so for example directly from a received message, not accumulated).

Comment (maqi) [reply src/section/shared_state.rs:195]: > so for example directly from a received message, not accumulated

yes.
So, basically, if the current mutation function has a `proof` passed it, it indicates it's `accumulated`, wo we don't need to check the signature validity anymore.
If no `proof` provided, then it maybe a direct message, then shall be checked. (however, maybe due to the serialisation/derialization, the checking place shall be in the calling place, not here.)

File: src/consensus/proof.rs [hunk 6]
```diff
@@ -6,6 +6,7 @@
 // KIND, either express or implied. Please review the Licences for the specific language governing
 // permissions and limitations relating to use of the SAFE Network Software.
 
+use serde::Serialize as SerdeSerialize;
```
Comment (madadam) [inline src/consensus/proof.rs:9]: Why not just `use serde::Serialize` ?

Comment (maqi) [reply src/consensus/proof.rs:9]: avoid confliction with std::Serialize

File: src/section/shared_state.rs [hunk 7]
```diff
@@ -52,25 +53,34 @@ impl SharedState {
...
-            );
+        if self.our_history.len() > 1 {
+            if *self != new {
+                error!(
+                    "shared state update - mismatch: old: {:?} --- new: {:?}",
+                    self, new
+                );
+                return Err(RoutingError::InvalidState);
+            }
+        } else if !new.self_verify() {
```
Comment (madadam) [inline src/section/shared_state.rs:70]: One more thing: verify that `self.our_history` and `new.our_history` have at least one key in common. Without it someone could in theory send us a completely fabricated `SharedState` which would still pass `self_verify`.

File: src/section/shared_state.rs [hunk 8]
```diff
@@ -52,25 +53,34 @@ impl SharedState {
...
 
         *self = new;
         self.handled_genesis_event = true;
+        Ok(())
+    }
+
+    fn self_verify(&self) -> bool {
+        self.our_history.self_verify()
+            && self.our_members.self_verify()
+            && self.sections.self_verify()
```
Comment (madadam) [inline src/section/shared_state.rs:83]: We are getting there :) But we should also check that all the proofs in all the containers are signed with keys that exists in `self.our_history`. WIth that and https://github.com/maidsafe/routing/pull/2155/files#r452245185, we should be good.

File: src/consensus/proof.rs (hunk 6 above)
Comment (madadam) [reply src/consensus/proof.rs:9]: What is `std::Serialize` ?

Comment (madadam) [reply src/consensus/proof.rs:9]: Do you mean `derive(Serialize)`? If yes, then that is not `std` but also `serde` and its a different namespace (trait vs proc-macro) so it doesn't conflict.

Comment (maqi) [reply src/consensus/proof.rs:9]: ah, my mistake, will change it

File: src/section/shared_state.rs (hunk 7 above)
Comment (maqi) [reply src/section/shared_state.rs:70]: ok. will do

File: src/section/shared_state.rs (hunk 8 above)
Comment (maqi) [reply src/section/shared_state.rs:83]: that will requires the key history to be pass down across couple of structs.  :(
will have a try

File: src/consensus/proof.rs [hunk 9]
```diff
@@ -75,20 +77,28 @@ impl Debug for ProofShare {
...
     pub proof: Proof,
 }
 
-impl<T> Proven<T> {
+impl<T: Serialize> Proven<T> {
     pub fn new(value: T, proof: Proof) -> Self {
         Self { value, proof }
     }
+
+    pub fn self_verify(&self, history: &SectionProofChain) -> bool {
```
Comment (madadam) [inline src/consensus/proof.rs:90]: Very minor point: maybe this should no longer be called `self_verify` as it does the trusted keys check too? Maybe call it just `verify`?.

**Comment by madadam** [comment]
Also maybe add some unit tests for the new logic? Could be done in a separate PR too.
